### PR TITLE
preAlpha2 multiclass

### PR DESCRIPTION
**Pull request by pabloT97**
Implement preAlpha2 multiclass and tests. 

There is an utils.py file with some helping functions for the model. My idea is that once the models are wrapped up we can have only one utils.py file common to all the models. For the moment I am keeping one utils file per model in order to avoid merging conflicts.

### CODE REVIEW

File: neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py [hunk 1]
```diff
@@ -1,250 +1,331 @@
...
+        if loss_function == None:
+            self.loss_function = self.cross_entropy_loss_wrapper()
+        else:
+            self.loss_function = loss_function
+        self.optimiser_args = optimiser_args
+        self.device = device
+        self.seed = seed
 
-        return sentences, labels
+    def cross_entropy_loss_wrapper(self)-> None:
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py:111]: Do you think you could unwrap this? Nested functions aren't a great thing.

File: neasqc_wp61/models/quantum/pre_alpha_2/utils.py
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/pre_alpha_2/utils.py:1]: Great!

File: tests/test_pre_alpha_2.py [hunk 2]
```diff
@@ -0,0 +1,180 @@
+import argparse 
+import os
+import random 
+import sys 
+import unittest
+
+import torch
+
+# The two lines below will be removed when converting the library to a package.
```
Comment (ryk-wolf) [inline tests/test_pre_alpha_2.py:9]: Thanks for noting this!

File: tests/test_pre_alpha_2.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+            cls.pre_alpha_2.preds_train,
+            cls.pre_alpha_2.preds_val,
+        ]
+        torch.manual_seed(args.seed)
+
+    def test_model_raises_error_if_not_enough_sentence_qubits(self)-> None:
+        """
+        Test that the model raises an error when thre are not enough sentence qubits.
+        """
+        with self.assertRaises(ValueError):
```
Comment (ryk-wolf) [inline tests/test_pre_alpha_2.py:55]: BEAUTIFUL!

File: tests/test_pre_alpha_2.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+        Test that the number of predictions and probabilities is equal
+        to the the number of epochs in traininig.
+        """
+        for probs,preds in zip(self.probabilites, self.predictions):
+            with self.subTest(probs=probs):
+                self.assertEqual(self.pre_alpha_2.epochs, len(probs))
+            with self.subTest(pred=preds):
+                self.assertEqual(self.pre_alpha_2.epochs, len(preds))
+    
+    def test_preds_are_integers(self):
```
Comment (ryk-wolf) [inline tests/test_pre_alpha_2.py:90]: Change the p1 p2 names to something a bit more explicit

File: tests/test_pre_alpha_2.py
Comment (ryk-wolf) [inline tests/test_pre_alpha_2.py:1]: The tests are getting better and better. Keep it up!

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
The testing has been improving A LOT! keep it up!

File: neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py [hunk 5]
```diff
@@ -1,250 +1,331 @@
...
-    A class to implement the lambeq version of pre-alpha
+    A class to implement sentence classification 
+    using lambeq library.
     """
-
-    @staticmethod
-    def load_dataset(
-        dataset : str  
-    ) -> list[list[str],list[list[int]]]:
+    # Define function dictionaries for ansatze and optimisers.
```
[the remark this answers is not in the document]
Comment (pabloT97) [reply neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py:19]: I assume you're referring to lines 20-33.  ansatz_functions_dict and optimisers_functions_dict are class variables, and that is why they come before init.

File: neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py (hunk 1 above)
Comment (pabloT97) [reply neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py:111]: No afaik. Lambeq requires the loss function to have two inputs (y_hat,y). For multiclass classification, the loss function will depend on the number of classes, which is an attribure of the class (see reshape_output_lambeq_model in line 130). That's why I've used the wrapper, so that loss function takes self as well as argument.

Of course if you know another way of doing this let me know. this was the best I could think of.

File: neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (ryk-wolf) [reply neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py:19]: I see, cheers!

File: neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py [hunk 6]
```diff
@@ -120,7 +128,7 @@ def cross_entropy_loss(y_hat,y):
         return cross_entropy_loss
     
     def reshape_output_lambeq_model(
-        self, x : torch.tensor, epsilon : float = 1e-13
+        self, tensor : torch.tensor, epsilon : float = 1e-13
```
Comment (ryk-wolf) [inline neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py:131]: Better, but if I'm correct this is a tensor of probabilities. If so, it should be called something like probabilities or probs. Tensor isn't really descriptive of what your object means, its a type.

File: tests/test_pre_alpha_2.py [hunk 7]
```diff
@@ -91,21 +91,43 @@ def test_preds_are_integers(self):
         """
         Test that the values predicted are integers.
         """
-        for preds in self.predictions:
-            for p1 in preds:
-                for p2 in p1:
-                    with self.subTest(p2 = p2):
-                        self.assertIs(type(p2), int)
+        preds_train_list = self.predictions[0]
```
Comment (ryk-wolf) [inline tests/test_pre_alpha_2.py:94]: very nice!

File: tests/test_pre_alpha_2.py [hunk 8]
```diff
@@ -91,21 +91,43 @@ def test_preds_are_integers(self):
...
     def test_probs_add_up_to_1(self):
         """
         Test that the probabilities predicted for each instance add up to one.
         """
-        for probs in self.probabilites:
-            for p1 in probs:
-                for p2 in p1:
-                    with self.subTest(p2 = p2):
-                        self.assertLessEqual(abs(sum(p2) -1), 1e-06)
+        probs_train_list = self.probabilities[0]
```
Comment (ryk-wolf) [inline tests/test_pre_alpha_2.py:113]: SAA :)

**Review by ryk-wolf — CHANGES_REQUESTED** [review]
Only one item left to change and we're good to merge!

File: neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pabloT97) [reply neasqc_wp61/models/quantum/pre_alpha_2/pre_alpha_2.py]: I have addressed this, I've realised it made no sense

**Review by ryk-wolf — APPROVED** [review]
